pkg/flow/internal/controller/component.go

```diff
@@ -0,0 +1,377 @@
+package controller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/flow/internal/dag"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/rfratto/gohcl"
+)
+
+// ComponentID is a fully-qualified name of a component. Each element in
+// ComponentID corresponds to a fragment of the period-delimited string;
+// "remote.http.example" is ComponentID{"remote", "http", "example"}.
+type ComponentID []string
+
+// BlockComponentID returns the ComponentID specified by an HCL block.
+func BlockComponentID(b *hcl.Block) ComponentID {
+	id := make(ComponentID, 0, 1+len(b.Labels)) // add 1 for the block type
+	id = append(id, b.Type)
+	id = append(id, b.Labels...)
+	return id
+}
+
+// String returns the string representation of a component ID.
+func (id ComponentID) String() string {
+	return strings.Join(id, ".")
+}
+
+// Equals returns true if id == other.
+func (id ComponentID) Equals(other ComponentID) bool {
+	if len(id) != len(other) {
+		return false
+	}
+	for i := 0; i < len(id); i++ {
+		if id[i] != other[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// ComponentOptions are used by ComponentNode to build managed components.
+type ComponentOptions struct {
+	Logger          log.Logger              // Logger shared between all managed components.
+	DataPath        string                  // Shared directory where component data may be stored
+	OnExportsChange func(cn *ComponentNode) // Invoked when the managed component updated its exports
+}
+
+// ComponentNode is a controller node which manages a user-defined component.
+//
+// ComponentNode manages the underlying component and caches its current
+// arguments and exports. ComponentNode manages the arguments for the component
+// from an HCL block.
+type ComponentNode struct {
+	id          ComponentID
+	nodeID      string // Cached from id.String() to avoid allocating new strings every time NodeID is called.
+	reg         component.Registration
+	opts        ComponentOptions
+	managedOpts component.Options
```

Comment:
These two made me do a double take, thought on embedding opts inside managedOpts?

Reply:
I removed opts to make it less confusing instead of combining the two structs 👍